### ISSUE TITLE
Metrics are reduced to a rank-local scalar before gather_for_metrics, biasing the final batch

### ISSUE THREAD

**Issue by behroozazarkhalili** [huggingface/trl#7010]
### Summary

Several metrics are reduced to a rank-local scalar before being handed to
`Accelerator.gather_for_metrics`. That defeats the de-duplication `gather_for_metrics` exists to do,
so on the final batch of an epoch the reported value is biased whenever the dataset does not divide
evenly across ranks.

`trl/trainer/reward_trainer.py` is affected, so this is not confined to `trl/experimental`.

### Why the pre-reduction breaks it

`gather_for_metrics` de-duplicates the padding accelerate added to the last batch by slicing the
gathered tensor along the batch dimension. From the installed accelerate source:

```python
elif self.gradient_state.remainder > 0:
    return tensor[: self.gradient_state.remainder]
```

That only works on a per-example tensor. A rank-local scalar has no batch dimension left to slice,
and the duplicated example is already folded into the number, so the truncation cannot remove it.

### The size of the error

Nine real examples, two ranks, local batch size five, so accelerate pads the second rank by one.
Exact rationals:

| what the padding repeats | reported | true | error |
| --- | --- | --- | --- |
| the first example | 18/5 = 3.6 | 4 | -0.4 |
| the last example | 22/5 = 4.4 | 4 | +0.4 |

A ten percent error on that batch, with the sign set by which element accelerate repeats. As a
control, with equal counts and no padding the mean of the rank means equals the true mean exactly, so
the bias comes specifically from the padded batch.

### The same file does it both ways

`trl/experimental/cpo/cpo_trainer.py`, four lines apart in one function:

```python
# correct: the per-example tensor reaches the de-duplicator, the mean comes after
self.accelerator.gather_for_metrics(policy_chosen_logps).detach().mean().item()

# biased: the mean happens first, so there is nothing left to de-duplicate
self.accelerator.gather_for_metrics(policy_rejected_logits.detach().mean()).mean().item()
```

That is what makes this look like an oversight rather than a deliberate convention.

### Sites

Found with an AST scan for a `.mean()` reaching `gather_for_metrics` or `gather_mean`, either inline
or through a variable assigned one line earlier. 21 sites in 7 files:

| file | sites |
| --- | --- |
| `trl/trainer/reward_trainer.py` | 775, 779 |
| `trl/experimental/xpo/xpo_trainer.py` | 387, 388, 394, 395, 411, 415 |
| `trl/experimental/online_dpo/online_dpo_trainer.py` | 1264, 1266, 1275, 1279 |
| `trl/experimental/ppo/ppo_trainer.py` | 905, 906, 908, 911 |
| `trl/experimental/cpo/cpo_trainer.py` | 906, 909 |
| `trl/experimental/orpo/orpo_trainer.py` | 864, 867 |
| `trl/experimental/tpo/tpo_trainer.py` | 708 |

`reward_trainer.py:774` is the clearest case: `(rewards_chosen > rewards_rejected).float().mean()`
reduces one boolean per pair to a scalar, then gathers it, so `accuracy` carries the duplicate.

I have not checked all 21 against their surrounding code. Anywhere the quantity is already global
rather than per-example, the ordering does not matter and the site can be dismissed.

### Suggested fix

Pass the per-example tensor and reduce after the gather, which is what `cpo_trainer.py:902` already
does:

```python
self.stats["logps/chosen"].append(gather_mean(chosen_model_logprobs))
```

### How I found it

A cross-model reviewer raised two of these sites on an unrelated PR. I confirmed the mechanism
against the installed accelerate source, computed the bias exactly, and swept the tree for the rest.
Happy to open the PR if the direction looks right.

**Comment by behroozazarkhalili** [comment]
Closing this myself. I found it by reading the code, not from a failure in a real training run, and per #6862 that isn't the bar here. Happy to reopen if someone hits it in practice.
